cmd/ctr/commands/tasks/migrate.go

```diff
@@ -0,0 +1,258 @@
+	ctx               context.Context
+	containerID       string
+	localClient       *containerd.Client
+	destClient        *containerd.Client
+}
+
+type ctrRemote struct {
+	ctx         context.Context
+	containerID string
+	client      taskApi.TasksClient
+}
+
+func (r *ctrRemote) StartIter() error {
+	_, err := r.client.StartIter(r.ctx, &taskApi.StartIterRequest{ContainerID: r.containerID})
+	return err
+}
+
+func (r *ctrRemote) StopIter() error {
+	_, err := r.client.StopIter(r.ctx, &taskApi.StopIterRequest{ContainerID: r.containerID})
+	return err
+}
+
+func (r *ctrRemote) DoRestore() error {
+	return nil
+}
+
+func (l *ctrLocal) PostDump() error {
+	return nil
+}
+
+func (l *ctrLocal) DumpCopyRestore(cr *criu.Criu, cfg phaul.PhaulConfig, last_cln_images_dir string) error {
+	task, err := getTask(l.ctx, l.localClient, l.containerID)
+	if err != nil {
+		return err
+	}
+	img, err := task.Checkpoint(l.ctx, func(r *containerd.CheckpointTaskInfo) error {
+		r.Options = &runctypes.CheckpointOptions{
+			PageServer: fmt.Sprintf("%s:%d", cfg.Addr, cfg.Port),
+			ParentPath: last_cln_images_dir,
+			Exit:       true,
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	if err := copyCheckpoint(l.ctx, l.localClient.ContentStore(), l.destClient.ContentStore(), img.Target().Digest); err != nil {
+		return err
+	}
+
+	newC, err := l.destClient.NewContainer(l.ctx, l.containerID, containerd.WithCheckpoint(img, l.containerID))
+	if err != nil {
+		return err
+	}
+
+	// FIXME: only support NullIO now
+	newTask, err := newC.NewTask(l.ctx, containerd.NullIO, containerd.WithTaskCheckpoint(img))
+	if err != nil {
+		return err
+	}
+
+	return newTask.Start(l.ctx)
+
+}
+
+func copyBlob(ctx context.Context, src content.Store, dest content.Store, digest digest.Digest) error {
+	ra, err := src.ReaderAt(ctx, digest)
+	if err != nil {
+		return err
+	}
+	defer ra.Close()
+	return content.WriteBlob(ctx, dest, digest.String(), content.NewReader(ra), 0, digest)
+}
+
+func copyCheckpoint(ctx context.Context, src content.Store, dest content.Store, digest digest.Digest) error {
+	if err := copyBlob(ctx, src, dest, digest); err != nil {
+		return err
+	}
+
+	b, err := content.ReadBlob(ctx, src, digest)
+	if err != nil {
+		return err
+	}
+	var index spec.Index
+	if err := json.Unmarshal(b, &index); err != nil {
+		return err
+	}
+
+	for _, m := range index.Manifests {
+		fmt.Printf("mediatype: %s\ndigest: %s\n", m.MediaType, m.Digest)
+		if err := copyBlob(ctx, src, dest, m.Digest); err != nil {
+			return err
+		}
+
+	}
+	return nil
+}
+
+func pullImage(ctx context.Context, src *containerd.Client, destAddr string, id string) error {
+	container, err := src.LoadContainer(ctx, id)
+	if err != nil {
+		return err
+	}
+	img, err := container.Image(ctx)
+	if err != nil {
+		return err
+	}
+	dest, err := containerd.New(destAddr)
+	if err != nil {
+		return err
+	}
+	defer dest.Close()
+	_, err = dest.Pull(ctx, img.Name(), containerd.WithPullUnpack)
```

Comment:
This is non-deterministic, because the registry may provide an updated image with the same name.

Also, IIUC, this PR does not work if the rootfs is modified.

Reply:
We can fix this by pulling image with digest specified:
``ctr images pull docker.io/library/alpine@sha256:d6bfc3baf615dc9618209a8d607ba2a8103d9c8a405b3bd8741d88b4bef36478
``


Yes, it only supports read-only rootfs.

Reply:
There is no guarantee that the image can be pulled.
If we are going to adopt this feature, we should use `Export()` and `Import()`

> Yes, it only supports read-only rootfs.

If the application is made to support read-only rootfs, the application is likely to made almost stateless, and unlikely to need live migration.
Do you have a specific usecase?

Reply:
In-memory data structure store, e.g memcached, redis and some message brokers.